File: src/com/cattailsw/nanidroid/install/NarStagedTree.java
package com.cattailsw.nanidroid.install;

import android.content.Context;
import java.util.List;

/**
 * Package-private capability boundary for a staged ghost tree.
 *
 * <p>The Context selects the app-owned staging root inside the backend. The
 * opaque handle and its validated inventory move together from {@link Tree}
 * to {@link Claim}. Neither type exposes a destination, path, token, stream,
 * or raw handle. D9b3 must treat process-death orphans as discard-only.
 */
final class NarStagedTree {
    private NarStagedTree() {}
    enum Error {
        OK, INVALID_OPTIONS, INVALID_TARGET, ROOT_TYPE, TARGET_TYPE, SYMLINK,
        SPECIAL_TYPE, INVALID_NAME, COMPONENT_LIMIT, PATH_LIMIT, DEPTH_LIMIT,
        ENTRY_COUNT_LIMIT, FILE_SIZE_LIMIT, TOTAL_SIZE_LIMIT, CYCLE,
        TREE_CHANGED, PERMISSION, RESOURCE, IO, VISITOR, CLOSE,
        INPUT, NATIVE, POLICY, CLOSED, CONSUMED, FOREIGN, WRONG_SESSION
    }
    interface Handle {}
    interface Backend {
        BeginResult begin(Context context,
                NarFilesystemInspector.TrustedRoot root,
                CharSequence target);
        NarStagedTreeInventory.Description describe(Handle handle);
        Error discard(Context context, Handle handle);
    }
    static final class BeginResult {
        private static final int ABSENT = 1;
        private static final int PRESENT = 2;
        private static final int FAILURE = 3;

        private final int kind;
        private final long storageDevice;
        private final long storageInode;
        private final Handle handle;
        private final Error primaryError;
        private final Error cleanupError;
        private BeginResult(int kind, long storageDevice,
                long storageInode, Handle handle,
                Error primaryError, Error cleanupError) {
            this.kind = kind;
            this.storageDevice = storageDevice;
            this.storageInode = storageInode;
            this.handle = handle;
            this.primaryError = primaryError;
            this.cleanupError = cleanupError;
        }
        static BeginResult absent(long storageDevice, long storageInode) {
            return new BeginResult(ABSENT, storageDevice, storageInode,
                    null, null, Error.OK);
        }
        static BeginResult present(Handle handle) {
            return new BeginResult(PRESENT, 0, 0, handle, null, Error.OK);
        }

        static BeginResult failure(
                Error primaryError, Error cleanupError, Handle handle) {
            return new BeginResult(FAILURE, 0, 0, handle,
                    primaryError, cleanupError);
        }
    }
    static final class Cleanup {
        private final Error nativeError;
        private Error discardError;
        private DiscardOwner recovery;

        private Cleanup(Error nativeError, Error discardError,
                DiscardOwner recovery) {
            this.nativeError = normalize(nativeError);
            this.discardError = normalize(discardError);
            this.recovery = recovery;
        }
        Error nativeError() { return nativeError; }
        Error discardError() { return discardError; }
        Error discard() {
            return recovery == null ? Error.OK : recovery.discard();
        }
    }
    static final class StageResult {
        private final Tree tree;
        private final Error error;
        private final Cleanup cleanup;
        private final String detail;

        private StageResult(Tree tree, Error error,
                Cleanup cleanup, String detail) {
            this.tree = tree;
            this.error = error;
            this.cleanup = cleanup;
            this.detail = detail == null ? "" : detail;
        }
        private static StageResult success(Tree tree) {
            return new StageResult(tree, Error.OK,
                    new Cleanup(Error.OK, Error.OK, null), "");
        }
        private static StageResult failure(
                Error error, Error nativeCleanup, String detail) {
            return new StageResult(null, normalizeFailure(error),
                    new Cleanup(nativeCleanup, Error.OK, null), detail);
        }
        boolean isSuccess() { return tree != null; }
        Tree tree() { return tree; }
        Error error() { return error; }
        Cleanup cleanup() { return cleanup; }
        String detail() { return detail; }
    }

    static final class Stager {
        private final Backend backend;

        Stager(Backend backend) {
            if (backend == null) throw new NullPointerException("backend");
            this.backend = backend;
        }

        Session session(Context context) {
            if (context == null) throw new NullPointerException("context");
            return new Session(this, context);
        }
    }

    static final class Session {
        private final Stager owner;
        private final Context context;

        private Session(Stager owner, Context context) {
            this.owner = owner;
            this.context = context;
        }

        StageResult stage(
                NarFilesystemInspector.TrustedRoot root, String target) {
            if (root == null || target == null) {
                return StageResult.failure(
                        Error.INPUT, Error.OK, "input");
            }
            Handle pending = null;
            DiscardOwner pendingOwner = null;
            StageResult result = null;
            try {
                BeginResult begun =
                        owner.backend.begin(context, root, target);
                if (begun == null) {
                    result = StageResult.failure(
                            Error.NATIVE, Error.OK, "begin");
                } else {
                    pending = begun.handle;
                    if (pending != null) {
                        pendingOwner = new DiscardOwner(
                                owner.backend, context, pending);
                    }
                    if (begun.kind == BeginResult.FAILURE) {
                        result = StageResult.failure(
                                begun.primaryError, begun.cleanupError,
                                "begin");
                    } else if (begun.kind == BeginResult.ABSENT
                            && pending == null) {
                        result = fromInventory(owner, this, context,
                                NarStagedTreeInventory.absent(
                                target, begun.storageDevice,
                                begun.storageInode), null);
                    } else if (begun.kind == BeginResult.PRESENT
                            && pending != null) {
                        NarStagedTreeInventory.Result inventory =
                                NarStagedTreeInventory.present(target,
                                        owner.backend.describe(pending));
                        result = fromInventory(owner, this, context,
                                inventory, pending);
                        if (result.isSuccess()) {
                            pending = null;
                            pendingOwner = null;
                        }
                    } else {
                        result = StageResult.failure(
                                Error.NATIVE, Error.OK, "begin shape");
                    }
                }
            } catch (RuntimeException error) {
                result = StageResult.failure(
                        Error.NATIVE, Error.OK, "backend");
            } catch (LinkageError error) {
                result = StageResult.failure(
                        Error.NATIVE, Error.OK, "backend");
            } finally {
                if (pendingOwner != null) {
                    if (result != null) {
                        result.cleanup.recovery = pendingOwner;
                    }
                    Error discarded;
                    try { discarded = pendingOwner.discard(); }
                    catch (OutOfMemoryError cleanup) { discarded = Error.NATIVE; }
                    if (result != null) {
                        result.cleanup.discardError = discarded;
                        if (discarded == Error.OK) {
                            result.cleanup.recovery = null;
                        }
                    }
                } else if (pending != null) {
                    try { Resource.discard(owner.backend, context, pending); }
                    catch (OutOfMemoryError cleanup) { /* preserve primary */ }
                }
            }
            return result;
        }

        ConsumeResult consume(Tree tree) {
            if (tree == null) return ConsumeResult.failure(Error.INPUT);
            synchronized (tree) {
                if (tree.owner != owner) {
                    return ConsumeResult.failure(Error.FOREIGN);
                }
                if (tree.session != this) {
                    return ConsumeResult.failure(Error.WRONG_SESSION);
                }
                if (tree.consumed) {
                    return ConsumeResult.failure(Error.CONSUMED);
                }
                if (tree.resource.isClosed()) {
                    return ConsumeResult.failure(Error.CLOSED);
                }
                Claim claim = new Claim(tree.resource);
                ConsumeResult transferred =
                        ConsumeResult.success(claim);
                tree.consumed = true;
                return transferred;
            }
        }
    }

    static final class Tree {
        private final Stager owner;
        private final Session session;
        private final Resource resource;
        private boolean consumed;

        private Tree(Stager owner, Session session, Resource resource) {
            this.owner = owner;
            this.session = session;
            this.resource = resource;
        }

        NarGhostTreePolicy.Manifest manifest() {
            return resource.inventory.manifest();
        }

        List<NarStagedTreeInventory.Entry> entries() {
            return resource.inventory.entries();
        }

        synchronized Error discard() {
            if (consumed) return Error.CONSUMED;
            return resource.discard();
        }
    }

    static final class Claim {
        private final Resource resource;

        private Claim(Resource resource) {
            this.resource = resource;
        }

        Error discard() {
            return resource.discard();
        }
    }

    static final class ConsumeResult {
        private final Claim claim;
        private final Error error;

        private ConsumeResult(Claim claim, Error error) {
            this.claim = claim;
            this.error = error;
        }

        private static ConsumeResult success(Claim claim) {
            return new ConsumeResult(claim, Error.OK);
        }

        private static ConsumeResult failure(Error error) {
            return new ConsumeResult(null, normalize(error));
        }

        boolean isSuccess() { return claim != null; }
        Claim claim() { return claim; }
        Error error() { return error; }
    }

    private static final class Resource {
        private final Backend backend;
        private final Context context;
        private final NarStagedTreeInventory.Result inventory;
        private Handle handle;
        private boolean closed;

        private Resource(
                Backend backend, Context context, Handle handle,
                NarStagedTreeInventory.Result inventory) {
            this.backend = backend;
            this.context = context;
            this.handle = handle;
            this.inventory = inventory;
            this.closed = false;
        }

        private synchronized boolean isClosed() {
            return closed;
        }

        private synchronized Error discard() {
            if (closed) return Error.OK;
            if (handle == null) {
                closed = true;
                return Error.OK;
            }
            Error result = discard(backend, context, handle);
            if (result == Error.OK) {
                handle = null;
                closed = true;
            }
            return result;
        }

        private static Error discard(
                Backend backend, Context context, Handle handle) {
            try {
                return normalize(backend.discard(context, handle));
            } catch (RuntimeException error) {
                return Error.NATIVE;
            } catch (LinkageError error) {
                return Error.NATIVE;
            }
        }
    }

    private static final class DiscardOwner {
        private final Backend backend;
        private final Context context;
        private Handle handle;

        private DiscardOwner(
                Backend backend, Context context, Handle handle) {
            this.backend = backend;
            this.context = context;
            this.handle = handle;
        }

        private synchronized Error discard() {
            if (handle == null) return Error.OK;
            Error result = Resource.discard(backend, context, handle);
            if (result == Error.OK) handle = null;
            return result;
        }
    }

    private static Error normalize(Error error) {
        return error == null ? Error.NATIVE : error;
    }
    private static Error normalizeFailure(Error error) {
        return error == null || error == Error.OK ? Error.NATIVE : error;
    }

    private static StageResult fromInventory(
            Stager owner, Session session, Context context,
            NarStagedTreeInventory.Result inventory, Handle handle) {
        if (!inventory.isSuccess()) {
            Error error = inventory.error()
                    == NarStagedTreeInventory.Error.POLICY
                    ? Error.POLICY : Error.NATIVE;
            return StageResult.failure(
                    error, Error.OK, inventory.detail());
        }
        Resource owned = new Resource(
                owner.backend, context, handle, inventory);
        return StageResult.success(new Tree(owner, session, owned));
    }
}

File: src/main/kotlin/com/cattailsw/nanidroid/install/NarStagedTree.kt
package com.cattailsw.nanidroid.install

import android.content.Context

/** Package-private capability boundary for a staged ghost tree. */
internal object NarStagedTree {
    private const val NATIVE_TOKEN_BYTES = 88
    enum class Error { OK, INVALID_OPTIONS, INVALID_TARGET, ROOT_TYPE, TARGET_TYPE, SYMLINK, SPECIAL_TYPE, INVALID_NAME, COMPONENT_LIMIT, PATH_LIMIT, DEPTH_LIMIT, ENTRY_COUNT_LIMIT, FILE_SIZE_LIMIT, TOTAL_SIZE_LIMIT, CYCLE, TREE_CHANGED, PERMISSION, RESOURCE, IO, VISITOR, CLOSE, INPUT, NATIVE, POLICY, CLOSED, CONSUMED, FOREIGN, WRONG_SESSION, BUSY }
    interface Handle
    interface Backend {
        fun begin(context: Context, root: NarFilesystemInspector.TrustedRoot, target: CharSequence): BeginResult?
        fun describe(handle: Handle): NarStagedTreeInventory.Description
        fun discard(context: Context, handle: Handle): Error?
    }
    private class NativeHandle(token: ByteArray?, val description: NarStagedTreeInventory.Description?) : Handle {
        val token: ByteArray
        init { require(token != null && token.size == NATIVE_TOKEN_BYTES && description != null) { "native handle" }; this.token = token.clone() }
    }
    class NativeBackend : Backend {
        private var loaded = false
        override fun begin(context: Context, root: NarFilesystemInspector.TrustedRoot, target: CharSequence): BeginResult? { ensureLoaded(); return nativeBegin(context.getDir("narfs-stage-v1", Context.MODE_PRIVATE).absolutePath, NarFilesystemInspector.sourceRootValue(root), target.toString()) }
        override fun describe(handle: Handle): NarStagedTreeInventory.Description = owned(handle).description!!
        override fun discard(context: Context, handle: Handle): Error? { ensureLoaded(); return fromNativeError(nativeDiscard(context.getDir("narfs-stage-v1", Context.MODE_PRIVATE).absolutePath, owned(handle).token)) }
        @Synchronized private fun ensureLoaded() { if (!loaded) { System.loadLibrary("narfs"); loaded = true } }
        private fun owned(handle: Handle): NativeHandle = handle as? NativeHandle ?: throw IllegalArgumentException("native handle")
    }
    class BeginResult private constructor(@get:JvmSynthetic internal val kind: Int, @get:JvmSynthetic internal val storageDevice: Long, @get:JvmSynthetic internal val storageInode: Long, @get:JvmSynthetic internal val handle: Handle?, @get:JvmSynthetic internal val primaryError: Error?, @get:JvmSynthetic internal val cleanupError: Error?) {
        companion object { private const val ABSENT=1; private const val PRESENT=2; private const val FAILURE=3
            @JvmStatic fun absent(device:Long,inode:Long)=BeginResult(ABSENT,device,inode,null,null,Error.OK)
            @JvmStatic fun present(handle:Handle)=BeginResult(PRESENT,0,0,handle,null,Error.OK)
            @JvmStatic fun failure(primary:Error?, cleanup:Error?, handle:Handle?)=BeginResult(FAILURE,0,0,handle,primary,cleanup)
        }
    }
    class Cleanup(private val nativeErrorValue: Error?, @get:JvmSynthetic @set:JvmSynthetic internal var discardError: Error?, @get:JvmSynthetic @set:JvmSynthetic internal var recovery: DiscardOwner?) {
        private val nativeError = normalize(nativeErrorValue); init { discardError=normalize(discardError) }
        fun nativeError()=nativeError; fun discardError()=discardError!!; fun discard()=recovery?.discard() ?: Error.OK
    }
    class StageResult private constructor(val tree: Tree?, val error: Error?, val cleanup: Cleanup, val detail: String?) {
        companion object { @JvmStatic fun success(tree:Tree)=StageResult(tree,Error.OK,Cleanup(Error.OK,Error.OK,null),""); @JvmStatic fun failure(error:Error?, cleanup:Error?, detail:String?)=StageResult(null,normalizeFailure(error),Cleanup(cleanup,Error.OK,null),detail ?: "") }
        fun isSuccess()=tree != null
    }
    class Stager(@get:JvmSynthetic internal val backend: Backend = NativeBackend()) { init { requireNotNull(backend) { "backend" } }; fun session(context:Context):Session { requireNotNull(context) { "context" }; return Session(this,context) } }
    class Session(private val owner:Stager, private val context:Context) {
        fun stage(root:NarFilesystemInspector.TrustedRoot?, target:String?):StageResult {
            if(root==null||target==null)return StageResult.failure(Error.INPUT,Error.OK,"input")
            var pending:Handle?=null; var pendingOwner:DiscardOwner?=null; var result:StageResult?=null
            try { val begun=owner.backend.begin(context,root,target); if(begun==null) result=StageResult.failure(Error.NATIVE,Error.OK,"begin") else { pending=begun.handle; if(pending!=null)pendingOwner=DiscardOwner(owner.backend,context,pending!!)
                result=when { begun.kind==1&&pending==null -> fromInventory(owner,this,context,NarStagedTreeInventory.absent(target,begun.storageDevice,begun.storageInode),null)
                    begun.kind==2&&pending!=null -> { val r=fromInventory(owner,this,context,NarStagedTreeInventory.present(target,owner.backend.describe(pending!!)),pending); if(r.isSuccess()){pending=null;pendingOwner=null};r }
                    begun.kind==3 -> StageResult.failure(begun.primaryError,begun.cleanupError,"begin")
                    else -> StageResult.failure(Error.NATIVE,Error.OK,"begin shape") }
            }} catch(_:RuntimeException){result=StageResult.failure(Error.NATIVE,Error.OK,"backend")} catch(_:LinkageError){result=StageResult.failure(Error.NATIVE,Error.OK,"backend")}
            finally { if(pendingOwner!=null){ if(result!=null)result!!.cleanup.recovery=pendingOwner; val discarded=try{pendingOwner!!.discard()}catch(_:OutOfMemoryError){Error.NATIVE}; if(result!=null){result!!.cleanup.discardError=discarded;if(discarded==Error.OK)result!!.cleanup.recovery=null} } else if(pending!=null) try{Resource.discard(owner.backend,context,pending!!)}catch(_:OutOfMemoryError){} }
            return result!!
        }
        fun consume(tree:Tree?):ConsumeResult { if(tree==null)return ConsumeResult.failure(Error.INPUT); synchronized(tree){ return when { tree.owner!==owner->ConsumeResult.failure(Error.FOREIGN); tree.session!==this->ConsumeResult.failure(Error.WRONG_SESSION); tree.consumed->ConsumeResult.failure(Error.CONSUMED); tree.resource.isClosed()->ConsumeResult.failure(Error.CLOSED); else->{tree.consumed=true;ConsumeResult.success(Claim(tree.resource))} } } }
    }
    class Tree internal constructor(@get:JvmSynthetic internal val owner:Stager, @get:JvmSynthetic internal val session:Session, @get:JvmSynthetic internal val resource:Resource) { @get:JvmSynthetic @set:JvmSynthetic internal var consumed=false; fun manifest()=resource.inventory.manifest()!!; fun entries()=resource.inventory.entries(); @Synchronized fun discard()=if(consumed)Error.CONSUMED else resource.discard() }
    class Claim(private val resource:Resource) { enum class State{READY,BUSY,CONSUMED}; private var state=State.READY; private var current:Lease?=null; private var directCleanup=false
        @Synchronized fun state()=state; @Synchronized fun lease():Lease? { if(state!=State.READY)return null; return Lease(this,resource).also{current=it;state=State.BUSY} }
        @Synchronized fun release(lease:Lease?):Error { if(lease==null||lease.owner!==this)return Error.FOREIGN;if(!lease.active||state==State.CONSUMED)return Error.CONSUMED;if(state!=State.BUSY||current!==lease)return Error.BUSY;lease.active=false;current=null;state=State.READY;return Error.OK }
        @Synchronized fun consume(lease:Lease?):Error { if(lease==null||lease.owner!==this)return Error.FOREIGN;if(!lease.active||state==State.CONSUMED)return Error.CONSUMED;if(state!=State.BUSY||current!==lease)return Error.BUSY;lease.active=false;lease.consumed=true;current=null;state=State.CONSUMED;return Error.OK }
        fun discard():Error { synchronized(this){if(state==State.BUSY)return Error.BUSY;if(state==State.READY){state=State.CONSUMED;directCleanup=true}else if(!directCleanup)return Error.CONSUMED};return resource.discard() }
        class Lease internal constructor(@get:JvmSynthetic internal val owner:Claim, private val resource:Resource) { @get:JvmSynthetic @set:JvmSynthetic internal var active=true; @get:JvmSynthetic @set:JvmSynthetic internal var consumed=false; fun manifest():NarGhostTreePolicy.Manifest { synchronized(owner){check(active&&owner.state==State.BUSY&&owner.current===this){"stale baseline lease"};return resource.inventory.manifest()!!} }; fun entries():List<NarStagedTreeInventory.Entry>{synchronized(owner){check(active&&owner.state==State.BUSY&&owner.current===this){"stale baseline lease"};return resource.inventory.entries()}}; fun discard()=synchronized(owner){if(!consumed)Error.CONSUMED else resource.discard()} }
    }
    class ConsumeResult private constructor(val claim:Claim?, val error:Error?) { companion object { @JvmStatic fun success(claim:Claim)=ConsumeResult(claim,Error.OK); @JvmStatic fun failure(error:Error?)=ConsumeResult(null,normalize(error)) }; fun isSuccess()=claim!=null }
    internal class Resource(val backend:Backend,val context:Context,var handle:Handle?,val inventory:NarStagedTreeInventory.Result) { private var closed=false; @Synchronized fun isClosed()=closed; @Synchronized fun discard():Error { if(closed)return Error.OK;if(handle==null){closed=true;return Error.OK};val r=discard(backend,context,handle!!);if(r==Error.OK){handle=null;closed=true};return r }; companion object { fun discard(backend:Backend,context:Context,handle:Handle)=try{normalize(backend.discard(context,handle))}catch(_:RuntimeException){Error.NATIVE}catch(_:LinkageError){Error.NATIVE} } }
    internal class DiscardOwner(private val backend:Backend,private val context:Context,private var handle:Handle?) { @Synchronized fun discard():Error { if(handle==null)return Error.OK;val r=Resource.discard(backend,context,handle!!);if(r==Error.OK)handle=null;return r } }
    private fun normalize(error:Error?)=error?:Error.NATIVE; private fun normalizeFailure(error:Error?)=if(error==null||error==Error.OK)Error.NATIVE else error
    @JvmStatic fun fromNativeBegin(stateCode:Int,errorCode:Int,cleanupCode:Int,storageDevice:Long,storageInode:Long,token:ByteArray?,paths:Array<String>?,types:IntArray?,sizes:LongArray?,ordinals:IntArray?,digests:ByteArray?):BeginResult { val error=fromNativeError(errorCode);val cleanup=fromNativeError(cleanupCode);val description=NarStagedTreeInventory.Description(storageDevice,storageInode,paths,types,sizes,ordinals,digests);val handle=if(token==null)null else NativeHandle(token,description);if(stateCode==1&&error==Error.OK&&cleanup==Error.OK&&handle==null&&paths?.isEmpty()==true&&types?.isEmpty()==true&&sizes?.isEmpty()==true&&ordinals?.isEmpty()==true&&digests?.isEmpty()==true)return BeginResult.absent(storageDevice,storageInode);if(stateCode==2&&error==Error.OK&&cleanup==Error.OK&&handle!=null)return BeginResult.present(handle);return BeginResult.failure(if(error==Error.OK)Error.NATIVE else error,cleanup,handle) }
    private fun fromNativeError(code:Int)=when { code in 0..20->Error.entries[code];code==100->Error.INPUT;else->Error.NATIVE }
    private fun fromInventory(owner:Stager,session:Session,context:Context,inventory:NarStagedTreeInventory.Result,handle:Handle?):StageResult { if(!inventory.isSuccess())return StageResult.failure(if(inventory.error()==NarStagedTreeInventory.Error.POLICY)Error.POLICY else Error.NATIVE,Error.OK,inventory.detail());return StageResult.success(Tree(owner,session,Resource(owner.backend,context,handle,inventory))) }
    @JvmStatic private external fun nativeBegin(stagingRoot:String,trustedRoot:String,target:String):BeginResult
    @JvmStatic private external fun nativeDiscard(stagingRoot:String,token:ByteArray):Int
}
